From f0364a7f3bbddc6f9a5fd62d84343c56817768f1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Alex=20R=C3=B8nne=20Petersen?= <alex@alexrp.com>
Date: Fri, 20 Feb 2026 22:14:35 +0100
Subject: [PATCH 130/133] std.heap.DebugAllocator: make BucketHeader.fromPage()
 use wrapping arithmetic
X-Developer-Signature: v=1; a=openpgp-sha256; l=1010; i=xtex@astrafall.org;
 h=from:subject; bh=iaHrTwReQpM2Xg1rbsPLAspVZGMaJVknoPda4AvfszQ=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78gR2cLx8cOr048ZZ9/5mn1Pqb9HbmbNoYrxs7vUJn
 7L/3LVr7ZjIwiDGxWAppshSZNjgzaqTzi+6rFwWZg4rE8gQaZEGBgYGBhYGvtzEvFIjHSM9U21D
 PUNDHSMdIwYuTgGY6mXhDP9L9sQL6WxzzvkmyxuTxa4540ziydd7jNcdDD257G2L6twjjAzdD6R
 Xpp9YfcB0YSl/wW1vlz7hM/fedp9+sVezMNGw/D4XAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

If we've allocated the very last page in the address space then these operations
will overflow and underflow respectively - which is fine.

(cherry picked from commit b5bcbf2a62aa513022fdb12df75699308f87b7bc)
---
 lib/std/heap/debug_allocator.zig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/std/heap/debug_allocator.zig b/lib/std/heap/debug_allocator.zig
index a8b88c16e07f..a895a4f570c9 100644
--- a/lib/std/heap/debug_allocator.zig
+++ b/lib/std/heap/debug_allocator.zig
@@ -285,7 +285,7 @@ pub fn DebugAllocator(comptime config: Config) type {
             canary: usize = config.canary,
 
             fn fromPage(page_addr: usize, slot_count: usize) *BucketHeader {
-                const unaligned = page_addr + page_size - bucketSize(slot_count);
+                const unaligned = page_addr +% page_size -% bucketSize(slot_count);
                 return @ptrFromInt(unaligned & ~(@as(usize, @alignOf(BucketHeader)) - 1));
             }
 
-- 
2.52.0

